homeassistant/components/media_player/sonos.py

```diff
@@ -43,16 +46,83 @@
 SUPPORT_SOURCE_TV = 'TV'
 SUPPORT_SOURCE_RADIO = 'Radio'
 
+SONOS_SCHEMA = vol.Schema({
+    ATTR_ENTITY_ID: cv.entity_ids,
+})
+
+# List of devices that have been registered
+DEVICES = []
+
 
 # pylint: disable=unused-argument, too-many-locals
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Setup the Sonos platform."""
     import soco
+    global DEVICES
+
+    def register_services():
```

Comment:
If you pass in `hass` as parameter, you can move this method outside of the `setup_platform(…)`. Will look a tad cleaner. You can also move the `_apply_service(…)`, `group_players_service(service)` etc out of the register services method.